sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchIterator.scala

```diff
@@ -107,22 +107,34 @@ class GpuColumnarBatchWithPartitionValuesIterator(
 
   private var leftValues: Array[InternalRow] = partValues
   private var leftRowNums: Array[Long] = partRowNums
+  private var nextBatchIter: Iterator[ColumnarBatch] = EmptyGpuColumnarBatchIterator
 
-  override def hasNext: Boolean = inputIter.hasNext
+  override def hasNext: Boolean = {
+    // Check if there is a next batch available, or if we need to process the current batch
+    if (nextBatchIter.hasNext) {
+      true
+    } else if (!inputIter.hasNext) {
+      false
+    } else {
+      val batch = inputIter.next()
+      nextBatchIter = if (partSchema.nonEmpty) {
+        val (readPartValues, readPartRows) = closeOnExcept(batch) { _ =>
+          computeValuesAndRowNumsForBatch(batch.numRows())
+        }
+        MultiFileReaderUtils.addMultiplePartitionValuesAndCloseIter(batch,
+          readPartValues, readPartRows, partSchema)
+      } else {
+        new SingleGpuColumnarBatchIterator(batch)
+      }
+      nextBatchIter.hasNext
```

Comment:
why did we move this code from `hasNext` to `next`?

Reply:
This fragment of code returns an `Iterator[ColumnarBatch]`. 
1. Iterator of multiple batches if splitting was required
2. Iterator of single batch if no split was required.

Hence, the `hasNext` condition should be governed by the `hasNext` of this new iterator.